storageccl: remove index id during rekey of primary index

madelynnblue opened the pull request:
During backup, spans are created based on their index, so the first
span for a table would be something like /Table/52/1. During restore,
we would restore these spans as they are during backup, resulting in a
span from /Table/51/<whatever> - /Table/52/1. The next time the split
queue is run it would ensure a table is split at the correct place
(/Table/52), leaving a forever-empty range at /Table/52 - /Table/52/1.

To fix this, teach the rekey method to recognize primary index start
spans and have it move those to the start of the table span.

Fixes #16317

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19052)
<!-- Reviewable:end -->

danhhz commented:
Please add some more detail to the commit message, including what the issue was, how it was happening, and why this fixes it

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/storageccl/key_rewriter.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvDG5SGN5fWAu1XxY:-KvgvDG5SGN5fWAu1XxZ:b-ttmqpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter.go#L207)):*
> ```Go
> 	}
> 	if b, id, idx, err := sqlbase.DecodeTableIDIndexID(newKey); err != nil {
> 		return roachpb.Span{}, errors.Wrapf(err, "could not rewrite key: %s", span.Key)
> ```

this probably wants a distinct error message so we can figure out which branch was taken if some customer hits this

---

*[pkg/ccl/storageccl/key_rewriter.go, line 211 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvPfGQ6ARVZM8ojYr:-KvgvPfGQ6ARVZM8ojYs:bfh049n) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter.go#L211)):*
> ```Go
> 		newKey = keys.MakeTablePrefix(uint32(id))
> 	}
> 	// Modify all spans that begin at the primary index to instead begin at the
> ```

This comments seems oddly placed. Move above the `sqlbase.DecodeTableIDIndexID` block?

---

*[pkg/ccl/storageccl/key_rewriter_test.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvY6OCB-eCICpnR9c:-KvgvY6OCB-eCICpnR9d:b-vvqidy) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter_test.go#L147)):*
> ```Go
> 	})
> 
> 	t.Run("span", func(t *testing.T) {
> ```

seems like we want a RESTORE level test, too? it's probably too easy for this to regress otherwise

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-:-KvgvcweYKAxJlIJBlAW:b-rw49wb)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Added commit log details.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/key_rewriter.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvDG5SGN5fWAu1XxY:-KvhfJutbX11XWUomk-j:bhr37ti) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter.go#L207)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this probably wants a distinct error message so we can figure out which branch was taken if some customer hits this
</blockquote></details>

We already duplicate the error message 4 other times. This one wraps the error from DecodeTableIDIndexID, which will be unique enough to tell it's from this line.

---

*[pkg/ccl/storageccl/key_rewriter.go, line 211 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvPfGQ6ARVZM8ojYr:-KvhfXfmKr-eseV_HlG-:b-a4wcvi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter.go#L211)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

This comments seems oddly placed. Move above the `sqlbase.DecodeTableIDIndexID` block?
</blockquote></details>

Oops. Done.

---

*[pkg/ccl/storageccl/key_rewriter_test.go, line 147 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvY6OCB-eCICpnR9c:-KvhnRIiOjHaWxKUKH79:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter_test.go#L147)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

seems like we want a RESTORE level test, too? it's probably too easy for this to regress otherwise
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-:-KvhrDkXWmA-s-tKNukN:b-bnqxuc)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 350 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-Kvi6FtfoOPDh3fkDww7:-Kvi6FtgHDspCcjv0o8E:b71ogfd) ([raw file](https://github.com/cockroachdb/cockroach/blob/1962e344cf6dd7480184de7fd24e888bb3491687/pkg/ccl/sqlccl/backup_test.go#L350)):*
> ```Go
> 		{
> 			var count int
> 			sqlDBRestore.QueryRow(`
> ```

it might be more maintainable to count the number of ranges that are part of the table span and assert that they're equal on the original and restored table

---

*[pkg/ccl/storageccl/key_rewriter.go, line 207 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-KvgvDG5SGN5fWAu1XxY:-Kvi6clFXfWps0FNPTxQ:bb7m07t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab57ec7d1ca9d85559c9132a0ad062fd3fa9aed/pkg/ccl/storageccl/key_rewriter.go#L207)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

We already duplicate the error message 4 other times. This one wraps the error from DecodeTableIDIndexID, which will be unique enough to tell it's from this line.
</blockquote></details>

okay

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-:-Kvi6fofCCAvO2vSbgKL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 350 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052#-Kvi6FtfoOPDh3fkDww7:-KviCIUe5ZzJQ5x1B9AB:bqa7dxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/1962e344cf6dd7480184de7fd24e888bb3491687/pkg/ccl/sqlccl/backup_test.go#L350)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

it might be more maintainable to count the number of ranges that are part of the table span and assert that they're equal on the original and restored table
</blockquote></details>

I don't think that would have caught this bug because the split queue doesn't run fast enough to make the new split which would trigger the test failure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19052)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
storageccl: remove index id during rekey of primary index

During backup, spans are created based on their index, so the first
span for a table would be something like /Table/52/1. During restore,
we would restore these spans as they are during backup, resulting in a
span from /Table/51/<whatever> - /Table/52/1. The next time the split
queue is run it would ensure a table is split at the correct place
(/Table/52), leaving a forever-empty range at /Table/52 - /Table/52/1.

To fix this, teach the rekey method to recognize primary index start
spans and have it move those to the start of the table span.

Fixes #16317
